### ISSUE TITLE
Support IEnumerable on ReadObject

### ISSUE THREAD

**Issue by micdenny** [Syroot/BinaryData#14]
Now it is possible to WriteObject every IEnumerable while you can't ReadObject for instance on List<T> this because under the wood the ReadObject creates an Array of T and then simply set on the instance property.

We could support every IEnumerable that has an implementation with a constructor that accept an array.

**Comment by RayKoopa** [comment]
Can you please provide an example of what is happening and what is expected?

**Comment by RayKoopa** [comment]
Might or might not be included in a 4.1.0

**Comment by micdenny** [comment]
> Might or might not be included in a 4.1.0

It could be important for someone but not for many, because I think when working with binary stream it's more common to see simple arrays and not generics.

So for me it's a go for `5.0.0`

**Comment by RayKoopa** [comment]
I'm actually not sure why I didn't implement it on first sight, probably because it isn't always given how to create an instance if arbitrary collections and adding it felt like a small hack. The serialization features have always been only an addition to the library rather than a primary concern.

**Comment by micdenny** [comment]
> The serialization features have always been only an addition to the library rather than a primary concern.

I agree, it could also be a separated package based on BinaryData.Core 😉

**Comment by RayKoopa** [comment]
Yeah, I thought about simply making it dependent upon Syroot.BinaryData (as it _is_ the core) with v5, but I have to implement all the features of 5 yet.

**Comment by micdenny** [comment]
we could think later for v6 where we can cut off the serialization and make it become a separated package

**Comment by RayKoopa** [comment]
Not sure about that. If at all I can perform the split in v5 aswell, it already breaks a lot of backwards compatibility due to new serialization attributes and renamed / shortened enumerations. I'll talk back to some other users about that too first.
